### PR TITLE
fixes #23369 - load vmware datastores async

### PR DESCRIPTION
**Pull request by timogoebel**
This commit introduces async loading of vSphere Datastores and Storage Pods.

This speeds up the time before the form is required and is a prerequisite for #3559, a feature requested in 2013 that we still haven't delivered.

foreman-vsphere-load-datastores-async-3

@ohadlevy: Do you mind reviewing this as this is react code?

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/common/forms/Select.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/forms/Select.js]: react 16.3 uses `createRef` API instead
Would you mind to use it?
https://reactjs.org/docs/refs-and-the-dom.html

File: webpack/assets/javascripts/react_app/redux/reducers/hosts/storage/vmware.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/reducers/hosts/storage/vmware.js]: instead of a lot of nested `.set` you can use `merge`

File: webpack/assets/javascripts/react_app/redux/actions/hosts/storage/vmware.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/actions/hosts/storage/vmware.js:38]: should it be exported while it's used only in this file?

File: webpack/assets/javascripts/react_app/redux/actions/hosts/storage/vmware.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/actions/hosts/storage/vmware.js:48]: ditto

File: webpack/assets/javascripts/react_app/components/hosts/storage/vmware/controller/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/hosts/storage/vmware/controller/index.js:24]: Could you add Proptype?  It would be much easier to understand the component's props with it

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @timogoebel  :+1: 
I left few inline comments about the javascript part :)

**Comment by timogoebel** [comment]
@amirfefer: Thanks for the review, will take a look at your invaluable comments (and probably learn a lot!). 👍

File: webpack/assets/javascripts/react_app/redux/actions/hosts/storage/vmware.js
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/hosts/storage/vmware.js:38]: on the other hand, i tend to prefer to call this action from the component's [componentDidMount](https://github.com/theforeman/foreman/blob/ae96cb70130968bdcc8b7a9a9af7edc7c6e0d9a8/webpack/assets/javascripts/react_app/components/hosts/storage/vmware/index.js#L26)

**Comment by timogoebel** [comment]
@amirfefer: I hope I addressed most of your comments. Unfortunately, I did not get `createRef` working. I guess there is enough refactoring in this PR already, we can always change that later.

**Comment by ares** [comment]
I've tested the functionality and works great with and without cache enabled. I defer to @amirfefer for the `createRef` pattern, but :+1: form me for manual testing. I haven't really read the code. Test failures are unrelated.

File: webpack/assets/javascripts/react_app/redux/reducers/hosts/storage/vmware.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/reducers/hosts/storage/vmware.js]: please use merge instead of multiple sets, its suppose to be faster :)

**Review by ohadlevy — COMMENTED** [review]
could you please also have a look at the storybook? I think it needs an update after these changes.

nice additions, thanks!

**Comment by timogoebel** [comment]
@ohadlevy: Sure, but after 20 minutes trying to get the storybook server running I gave up. Does it work for you?

```
$ npm run storybook

> TheForemanDevDeps@1.20.0 storybook /Users/f0218175/foremandev/foreman
> start-storybook -p 6006

@storybook/react v3.2.13

=> Loading custom .babelrc
=> Loading custom addons config.
=> Loading custom webpack config (full-control mode).
/Users/f0218175/foremandev/foreman/.storybook/webpack.config.js:7
  defaultConfig.module.rules = [
                ^

TypeError: Cannot read property 'module' of undefined
    at module.exports (/Users/f0218175/foremandev/foreman/.storybook/webpack.config.js:7:17)
    at exports.default (/Users/f0218175/foremandev/foreman/node_modules/@storybook/react/dist/server/config.js:62:12)
    at exports.default (/Users/f0218175/foremandev/foreman/node_modules/@storybook/react/dist/server/middleware.js:19:37)
    at Object.<anonymous> (/Users/f0218175/foremandev/foreman/node_modules/@storybook/react/dist/server/index.js:153:34)
    at Module._compile (module.js:660:30)
    at Object.Module._extensions..js (module.js:671:10)
    at Module.load (module.js:573:32)
    at tryModuleLoad (module.js:513:12)
    at Function.Module._load (module.js:505:3)
    at Module.require (module.js:604:17)
```

**Comment by ohadlevy** [comment]
@timogoebel it seems to work for me:
```
Switched to a new branch '6103'
➜  foreman git:(6103) npm run storybook                                                                                                                     git:(6103|) 

> TheForemanDevDeps@1.20.0 storybook /home/ohad/git/foreman
> start-storybook -p 6006

info @storybook/react v3.4.8
info 
info => Loading custom .babelrc
info => Loading custom addons config.
info => Loading custom webpack config (full-control mode).
 66% building modules 623/657 modules 34 active ...oreman/node_modules/lodash/isArray.jsDEPRECATION WARNING on line 1, column 8 of stdin:
Including .css files with @import is non-standard behaviour which will be removed in future versions of LibSass.
Use a custom importer to maintain this behaviour. Check your implementations documentation on how to create a custom importer.

webpack built d4fc339334bb701f4e07 in 13336ms                                             
info Storybook started on => http://localhost:6006/
```

**Comment by ares** [comment]
I have the similar issue without this PR, the npm command starts the server but navigating to storybook, I see `Cannot read property 'defaults' of undefined` 
story

File: webpack/assets/javascripts/react_app/components/hosts/storage/vmware/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/hosts/storage/vmware/index.js]: PropTypes.object is not informative 
`Proptype.shape` is much better for arrays and objects.
Plus, if a prop is required, add `isRequired` to it
i.e:
```javascript
PropTypes.shape({
    color: PropTypes.string,
    fontSize: PropTypes.number
  }).isRequired, // if a prop is required, add `.isRequired`, else just add a default value
```

File: webpack/assets/javascripts/react_app/components/hosts/storage/vmware/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/hosts/storage/vmware/index.js]: you can use default values with Proptype:
```javascript
StorageContainer.defaultProps = {
  volumes: []
};
```

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @timogoebel :+1: 
just few minor issues with `PropType`

**Comment by ohadlevy** [comment]
what is the status of this PR? (if its ready for today's cut of date?)

**Comment by timogoebel** [comment]
@ohadlevy: I won't be able to look at this today. Let's to it in the 1.21 cycle.

**Comment by timogoebel** [comment]
@amirfefer, @ohadlevy, @ares: Improved the prop types and fixed up storybook. Can you please re-review?

**Review by amirfefer — DISMISSED** [review]
JS part looks good , thanks @timogoebel :+1: 
I did not test it though (have no vmware)

**Comment by amirfefer** [comment]
regardless of this PR,  I'd like to mention that i'm not a fan of spinners, especially when it comes on inputs, because it might lead to too many spinners in a page.  There's a much better solution IMO - skeleton loading:
for example:

skeleton

@theforeman/ui-ux, @Rohoover  any thoughts ?

**Comment by timogoebel** [comment]
@amirfefer: I fully agree, but that's definitely out of the scope of this PR.

**Comment by terezanovotna** [comment]
Love the idea of skeleton loading! [Opened a design issue for it. ](https://github.com/patternfly/patternfly-design/issues/726)

Can we make the button next to Disk Name look more like [patternfly button?](https://www.patternfly.org/pattern-library/widgets/#buttons)

**Comment by timogoebel** [comment]
> Can we make the button next to Disk Name look more like [patternfly button?](https://www.patternfly.org/pattern-library/widgets/#buttons)

We can, but not in this PR. :-)

**Comment by Rohoover** [comment]
I like the skeleton as well but we would need to get this into PF since this is a departure.

**Comment by timogoebel** [comment]
@amirfefer, @ares: Anything missing here?

**Comment by amirfefer** [comment]
LGFM :+1:

**Comment by timogoebel** [comment]
This is my final attempt to get this in, I rebased the code and added support for statistics of the datastore and storage pods.

**Review by ohadlevy — DISMISSED** [review]
LGTM, I don't have a working VMWare env to test.

**Comment by lizagilman** [comment]
@timogoebel or anyone interested, regarding storybook issue (if you still have it), this workaround works for me - https://gist.github.com/lizagilman/d3a94114fc3fd3214138ec120fca4a5c

**Comment by timogoebel** [comment]
@lizagilman: Thanks. I chose the "reboot" approach. That was quite popular in ancient times.
`rm -rf node_module && npm install`.

**Comment by timogoebel** [comment]
@chris1984: Can you do me a favor and test and ack this? Thanks a bunch.

**Comment by chris1984** [comment]
> @chris1984: Can you do me a favor and test and ack this? Thanks a bunch.

Sure thing :)

**Comment by amirfefer** [comment]
I've test it with vmware env and it works for me:
vmware

Looks like a snapshot needs an update

**Review by amirfefer — DISMISSED** [review]

**Review by chris1984 — DISMISSED** [review]
ACK, looks good Thanks @timogoebel

**Comment by chris1984** [comment]
[test foreman]

**Comment by ares** [comment]
It seems like test failures are related, mind to take a look?

**Comment by timogoebel** [comment]
Fixed the tests.

**Comment by timogoebel** [comment]
Tests passed.

File: package.json
Comment (ohadlevy) [inline package.json:86]: I tried packaging this, but it fails, I assume before merge we would need a packaging ACK.

EDIT: for completion, the error is
```
./add_npm_package.sh number_helpers 0.1.1
Found 0 dependencies - using single strategy
Making directory...FINISHED
Creating specs and downloading sources...---- npm2rpm ----
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
 - Starting npm module download: https://registry.npmjs.org/number_helpers/-/number_helpers-0.1.1.tgz
 - Unpacking in /tmp/npm2rpm-15831E09pZQE4YSYD ...
 - Finished extracting for number_helpers
 - Reading package.json for number_helpers
Warning: No README data
Warning: No license field.
 - Finished reading package.json for number_helpers
fs.js:133
  throw new ERR_INVALID_CALLBACK();
  ^

TypeError [ERR_INVALID_CALLBACK]: Callback must be a function
    at maybeCallback (fs.js:133:9)
    at Object.writeFile (fs.js:1139:14)
    at writeSpecFile (/home/ohad/git/foreman-packaging/node_modules/npm2rpm/bin/npm2rpm.js:87:6)
    at Extract.tar_extract.stream.on (/home/ohad/git/foreman-packaging/node_modules/npm2rpm/bin/npm2rpm.js:78:5)
    at Extract.emit (events.js:187:15)
    at DirWriter.<anonymous> (/home/ohad/git/foreman-packaging/node_modules/tar/lib/extract.js:81:8)
    at DirWriter.emit (events.js:187:15)
    at end (/home/ohad/git/foreman-packaging/node_modules/fstream/lib/writer.js:324:14)
    at /home/ohad/git/foreman-packaging/node_modules/fstream/lib/writer.js:314:34
    at endUtimes (/home/ohad/git/foreman-packaging/node_modules/fstream/lib/writer.js:238:46)
```

Comment (ekohl) [reply package.json:86]: I think that was a warning for a while already but likely that there's an update which makes it strict. I'll have a look.

Comment (ohadlevy) [reply package.json:86]: also TODO: add this package to webpack vendor file (can be done in a follow up pr).

Comment (ekohl) [reply package.json:86]: https://github.com/theforeman/foreman-packaging/pull/3174. Luckily in https://github.com/dLobatog/npm2rpm/pull/48 I already fixed that issue.

Comment (ohadlevy) [reply package.json:86]: @ekohl great, once you are ready please ack this PR.

**Review by ekohl — APPROVED** [review]
Packaging ACK, haven't looked at the code :)

**Review by chris1984 — APPROVED** [review]

**Comment by ohadlevy** [comment]
thanks @timogoebel!

**Comment by timogoebel** [comment]
Thanks, @ohadlevy + @amirfefer + @ares + @ekohl + @chris1984!

**Pull request merged**
